### PR TITLE
Sector spawn system

### PR DESCRIPTION
**Pull request by FuJa0815**
**Brief Description of What This PR Does**

Overhauls the spawn system to utilize sector spawning.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [ ] Initial code review passed (this and further items should not be checked by the PR author)
- [ ] Functionality is confirmed working by another person (see above checklist link)
- [ ] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author.

### CODE REVIEW

**Comment by FuJa0815** [comment]
I think the first iteration of the technical implementation should be more or less done. Balancing will be horrible in this state.

File: src/general/Sector.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        return $"{X};{Y}";
+    }
+
+    private static (int X, int Y) GetSectorCoords(Vector2 position)
+    {
+        var x = (int)(position.x / Constants.SECTOR_SIZE);
+        var y = (int)(position.y / Constants.SECTOR_SIZE);
+        return (x, y);
+    }
```
Comment (Maxonovien) [inline src/general/Sector.cs]: Sector size should probably be factored into a parameter for the sector, seeing that this class could be used for other purposes, especially in cloud systems (which uses a sector logic without mentioning it)

File: src/microbe_stage/Membrane.cs [hunk 2]
```diff
@@ -269,10 +271,11 @@ public Vector2 FindClosestOrganelles(Vector2 target)
     {
         // The distance we want the membrane to be from the organelles squared.
         float closestSoFar = 4;
-        Vector2 closest = new Vector2(INVALID_FOUND_ORGANELLE, INVALID_FOUND_ORGANELLE);
+        Vector2 closest = invalidVector;
 
-        foreach (var pos in OrganellePositions)
+        for (var i = 0; i < OrganellePositions.Length; i++)
         {
+            var pos = OrganellePositions[i];
```
Comment (Maxonovien) [inline src/microbe_stage/Membrane.cs]: Why did you change this? The previous handling looked better in my eyes.

More generally I feel you changed quite a lot of files (sometimes with good reasons) with no relation to what you wanted to bring, and I feel this makes this PR very dense...

Comment (FuJa0815) [reply src/microbe_stage/Membrane.cs]: This change was one of many tiny performance improvements I've tried to spawning microbes

Comment (hhyyrylainen) [reply src/microbe_stage/Membrane.cs]: A separate PR is probably warranted as this doesn't fundamentally have anything to do with spawn sectors...

File: src/microbe_stage/PatchManager.cs [hunk 3]
```diff
@@ -156,60 +152,40 @@ private void HandleCellSpawns(Patch patch)
         foreach (var entry in patch.SpeciesInPatch)
         {
             var species = entry.Key;
+            var populationHere = entry.Value;
```
Comment (Maxonovien) [inline src/microbe_stage/PatchManager.cs:155]: Two concerns on this:
a) wouldn't it be more efficient not to instantiate variables here
b) if we stick by this, we probably should settle for a solution when dealing with dictionary entries, cause other parts in the code do not create such variables.

And, if clarity is the issue, perhaps a more telling name than entry could be considered.

File: src/microbe_stage/SpawnedRigidBody.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿using Godot;
+
+/// <summary>
+///   All nodes that can be spawned with the spawn system must derivative from this class
```
Comment (Maxonovien) [inline src/microbe_stage/SpawnedRigidBody.cs:4]:
```suggestion
///   All nodes that can be spawned with the spawn system must be a derivative from this class
```

And perhaps we'd even prefer another term like "must inherit this class"?

File: src/microbe_stage/Spawner.cs [hunk 5]
```diff
@@ -1,50 +1,113 @@
...
 
 /// <summary>
 ///   Spawner that can be added to a SpawnSystem to be used for spawning things
 /// </summary>
 public abstract class Spawner
 {
+    private float[]? binomialValuesCache;
+
+    public virtual int BinomialN => 0;
+    public virtual float BinomialP => 0;
```
Comment (Maxonovien) [inline src/microbe_stage/Spawner.cs:14]: Ok, there is the explanation for the factorial args. However, would we not want to swap that to an actual utils class such as the math one?

File: src/microbe_stage/Spawner.cs [hunk 6]
```diff
@@ -1,50 +1,113 @@
...
     /// </summary>
-    /// <param name="worldNode">The parent node of spawned entities</param>
-    /// <param name="location">Location the spawn system wants to spawn a thing at</param>
-    /// <returns>An enumerator that on each next call spawns one thing</returns>
-    public abstract IEnumerable<ISpawned>? Spawn(Node worldNode, Vector3 location);
+    /// <returns>Returns the amount of spawns in a sector.</returns>
+    protected virtual int GetSpawnsInASector(float sectorDensity, Random random)
+    {
+        if (BinomialP == 0)
+            return 0;
```
Comment (Maxonovien) [inline src/microbe_stage/Spawner.cs:91]: What is binomial P supposed to stand for?

File: src/microbe_stage/Spawner.cs [hunk 7]
```diff
@@ -1,50 +1,113 @@
...
+
+        var nextRandom = random.NextFloat() * sectorDensity;
+        var binomialSum = 0f;
+        var i = 0;
+        do
+        {
+            binomialSum += BinomialValues[i++];
+        }
+        while (nextRandom >= binomialSum);
```
Comment (Maxonovien) [inline src/microbe_stage/Spawner.cs:100]: Why not using a simple while loop here?

File: src/microbe_stage/Spawner.cs [hunk 8]
```diff
@@ -1,50 +1,113 @@
...
+
+        return i - 1;
+    }
```
Comment (Maxonovien) [inline src/microbe_stage/Spawner.cs:103]: Why this probability law in particular? I assume it at least privileges central values, since nCp is highest for p = n/2 (if I'm not mistaken).

If it is a known method, could you please add a reference in the doc? That would be super helpful in reviewing the mechanics, especially if we consider tweaking it.

File: src/microbe_stage/Spawners.cs [hunk 9]
```diff
@@ -366,7 +364,11 @@ public CompoundCloudSpawner(Compound compound, CompoundCloudSystem clouds, float
         this.amount = amount;
     }
 
-    public override IEnumerable<ISpawned>? Spawn(Node worldNode, Vector3 location)
+    public override int BinomialN => (int)(amount / 50000);
+    public override float BinomialP => 0.1f;
```
Comment (Maxonovien) [inline src/microbe_stage/Spawners.cs:368]: I'm definitely confused as to what these values represent in your algorithm, and why you give the these values.

**Review by Maxonovien — COMMENTED** [review]
Some general comments about the maths and a few minor remarks. Overall it seems well thought, but I think some parts remain to obscure in their details.  Keep up the great work!

File: src/microbe_stage/PatchManager.cs (hunk 3 above)
Comment (hhyyrylainen) [reply src/microbe_stage/PatchManager.cs:155]: I've recently discovered that you can even destructure variables (usually) with something like: `foreach(var (thing, secondThing) in container)` which is very awesome. I would expect the compiler to optimize away any temporary variables or at least that isn't much of a problem if this is code that isn't ran every frame.

**Comment by RacerBG** [comment]
This PR should be superseded already by PR #3235 .

**Comment by hhyyrylainen** [comment]
This has bunch of stuff like storing the spawned in sector data etc. that this is actually even more advanced than that spawn system change that did get done.
